livekit/src/room/metrics/mod.rs

```diff
@@ -0,0 +1,382 @@
+// Copyright 2025 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::time::Duration;
+
+use livekit_protocol::{self as proto, data_packet};
+use proto::{DataPacket, MetricLabel, MetricSample, MetricsBatch, TimeSeriesMetric};
+
+use crate::prelude::LocalTrackPublication;
+use crate::room::{Room, RtcEngine};
+use libwebrtc::stats::{InboundRtpStats, MediaSourceStats, OutboundRtpStats, RtcStats};
+use tokio_util::sync::CancellationToken;
+
+use super::id::{ParticipantIdentity, TrackSid};
+use super::DataPacketKind;
+
+type GenericMetricLabel<T> = (MetricLabel, T);
+
+const QUALITY_LIMITATIONS: [GenericMetricLabel<&str>; 3] = [
+    (MetricLabel::ClientVideoPublisherQualityLimitationDurationCpu, "cpu"),
+    (MetricLabel::ClientVideoPublisherQualityLimitationDurationBandwidth, "bandwidth"),
+    (MetricLabel::ClientVideoPublisherQualityLimitationDurationOther, "other"),
+];
+
+pub struct RtcMetricsManager {}
+
+impl RtcMetricsManager {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn micro_to_milli(micros: i64) -> i64 {
+        micros / 1000
+    }
+
+    fn get_or_create_index(&self, strings: &mut Vec<String>, string: String) -> u32 {
+        let position: Option<usize> = strings.iter().position(|s| s == &string);
+
+        match position {
+            Some(index) => return (index as u32) + (MetricLabel::PredefinedMaxValue as u32),
+            None => {
+                strings.push(string);
+                return ((strings.len() - 1) as u32) + (MetricLabel::PredefinedMaxValue as u32);
+            }
+        }
+    }
+
+    fn create_metric_sample(&self, timestamp_ms: i64, value: f32) -> MetricSample {
+        MetricSample { timestamp_ms, value, ..Default::default() }
+    }
+
+    fn create_time_series(
+        &self,
+        label: MetricLabel,
+        strings: &mut Vec<String>,
```

Comment:
Any reason why this need to be a mutable reference? It seems like the "caller" doesn't expect it? So maybe it's fine to clone inside this method and keep the modifications in this scope

Reply:
Yes actually:

- Strings are modified here: https://github.com/livekit/rust-sdks/blob/fd33fe818a4e57fa69a6297e0bd96866721d6e88/livekit/src/room/metrics/mod.rs#L54
- but are created here: https://github.com/livekit/rust-sdks/blob/fd33fe818a4e57fa69a6297e0bd96866721d6e88/livekit/src/room/metrics/mod.rs#L230